pycyphal/dsdl/_import_hook.py

```diff
@@ -0,0 +1,149 @@
+        _logger.debug("output dir: %s", output_directory)
+
+        self.lookup_directories = list(map(str, lookup_directories))
+        self.output_directory = output_directory
+        self.allow_unregulated_fixed_port_id = allow_unregulated_fixed_port_id
+        self.root_namespace_directories: Iterable[pathlib.Path] = list()
+
+        # Build a list of root namespace directories from lookup directories.
+        # Any dir inside any of the lookup directories is considered a root namespace if it matches regex
+        for dir in self.lookup_directories:
+            for namespace in pathlib.Path(dir).iterdir():
+                if namespace.is_dir() and re.match(r"[a-zA-Z_][a-zA-Z0-9_]*", namespace.name):
+                    _logger.debug("Using root namespace %s at %s", namespace.name, namespace)
+                    self.root_namespace_directories.append(namespace)
+
+    def find_source_dir(self, root_namespace: str) -> Optional[pathlib.Path]:
+        """
+        Finds DSDL source directory for a given root namespace name.
+        """
+        for namespace_dir in self.root_namespace_directories:
+            if namespace_dir.name == root_namespace:
+                return namespace_dir
+        return None
+
+    def is_compiled(self, root_namespace: str) -> bool:
+        """
+        Returns true if given root namespace exists in output directory (compiled).
+        """
+        return pathlib.Path(self.output_directory, root_namespace).exists()
+
+    def find_spec(self, fullname, path, target=None):
+        _logger.debug("Attempting to load module %s as DSDL", fullname)
+
+        # Translate module name to DSDL root namespace
+        root_namespace = root_namespace_from_module_name(fullname)
+
+        root_namespace_dir = self.find_source_dir(root_namespace)
+        if not root_namespace_dir:
+            return None
+
+        _logger.debug("Found root namespace %s in DSDL source directory %s", root_namespace, root_namespace_dir)
+
+        if not self.is_compiled(root_namespace):
+            _logger.debug("Compiling DSDL in %s", root_namespace_dir)
+            compile(
+                root_namespace_dir,
+                self.root_namespace_directories,
+                self.output_directory,
+                self.allow_unregulated_fixed_port_id,
+            )
+
+        compiled_module_dir = pathlib.Path(self.output_directory, root_namespace)
+        module_location = compiled_module_dir.joinpath("__init__.py")
+        submodule_locations = [str(compiled_module_dir)]
+
+        return spec_from_file_location(fullname, module_location, submodule_search_locations=submodule_locations)
+
+
+def get_default_lookup_dirs():
+    return os.environ.get("CYPHAL_PATH", "").replace(os.pathsep, ";").split(";")
+
+
+def get_default_output_dir():
+    pycyphal_dir = os.environ.get("PYCYPHAL_PATH", pathlib.Path.home().joinpath(".pycyphal"))
+    return pathlib.Path(pycyphal_dir, "compiled")
```

Comment:
Why do we need this one extra directory here? Why not simply `~/.pycyphal`? As a user, I would expect a PATH directory to point directly to the resource of interest.

Reply:
My thinking was that `PYCYPHAL_PATH` refers to general directory containing everything related to pycyphal. Maybe in the future it could contain some config files, cache, register database, whatever. If we don't include additional `compiled` directory then it becomes ambiguous what is what. So it's kind of future insurance.

If directory is to only compiled DSDL's then maybe env var should be `PYCYPHAL_COMPILED_PATH`? Anyway, let me know what you think.

Reply:
>Maybe in the future it could contain some config files, cache, register database, whatever.

I think that would necessitate a separate variable that doesn't need to be called `*_PATH`. What I think is crucial is that `PYCYPHAL_PATH` points directly to the directory where the compiled modules are located, otherwise it's a usability disaster.